plugins/vpp/abfplugin/descriptor/abf.go

```diff
@@ -0,0 +1,281 @@
+}
+
+// Validate validates VPP ABF configuration.
+func (d *ABFDescriptor) Validate(key string, abfData *abf.ABF) error {
+	if abfData.Index == "" {
+		return api.NewInvalidValueError(ErrABFInvalidIndex, "index")
+	}
+	_, err := strconv.Atoi(abfData.Index)
+	if err != nil {
+		return api.NewInvalidValueError(ErrABFInvalidIndex, "index")
+	}
+	if abfData.AclName == "" {
+		return api.NewInvalidValueError(ErrABFWithoutACL, "acl_name")
+	}
+	return nil
+}
+
+// Create validates ABF (mainly index), verifies ACL existence and configures ABF policy. Attached interfaces
+// are put to metadata together with the ABF index to make it available for other ABF descriptors.
+func (d *ABFDescriptor) Create(key string, abfData *abf.ABF) (*abfidx.ABFMetadata, error) {
+	abfIdx, err := getABFIndex(abfData)
+	if err != nil {
+		d.log.Error(err)
+		return nil, err
+	}
+
+	// get ACL index
+	aclData, exists := d.aclIndex.LookupByName(abfData.AclName)
+	if !exists {
+		err := errors.Errorf("failed to obtain metadata for ACL %s", abfData.AclName)
+		d.log.Error(err)
+		return nil, err
+	}
+
+	// add new ABF policy
+	if err := d.abfHandler.AddAbfPolicy(uint32(abfIdx), aclData.Index, abfData.ForwardingPaths); err != nil {
+		d.log.Error(err)
+		return nil, err
+	}
+
+	// fill the metadata
+	metadata := &abfidx.ABFMetadata{
+		Index:    uint32(abfIdx),
+		Attached: abfData.AttachedInterfaces,
+	}
+
+	return metadata, nil
+}
+
+// Delete removes ABF policy
+func (d *ABFDescriptor) Delete(key string, abfData *abf.ABF, metadata *abfidx.ABFMetadata) error {
+	// ACL ID is not required
+	return d.abfHandler.DeleteAbfPolicy(metadata.Index, abfData.ForwardingPaths)
+}
+
+// UpdateWithRecreate is always set to true since there is no binary API to specialy handle a part
+// of the config - the whole ABF policy needs to be removed and created.
+func (d *ABFDescriptor) UpdateWithRecreate(key string, oldAbfData, newAbfData *abf.ABF, oldMetadata *abfidx.ABFMetadata) bool {
+	// always recreate
+	return true
+}
+
+// Retrieve returns ABF policies from the VPP.
+func (d *ABFDescriptor) Retrieve(correlate []adapter.ABFKVWithMetadata) (abfs []adapter.ABFKVWithMetadata, err error) {
+	// Retrieve VPP configuration.
+	abfPolicies, err := d.abfHandler.DumpABFPolicy()
+	if err != nil {
+		return nil, errors.Errorf("failed to dump ABF policy: %v", err)
+	}
+
+	for _, abfPolicy := range abfPolicies {
+		abfs = append(abfs, adapter.ABFKVWithMetadata{
+			Key:   abf.Key(abfPolicy.ABF.Index),
+			Value: abfPolicy.ABF,
+			Metadata: &abfidx.ABFMetadata{
+				Index:    abfPolicy.Meta.PolicyID,
+				Attached: abfPolicy.ABF.AttachedInterfaces,
+			},
+			Origin: api.FromNB,
+		})
+	}
+
+	return abfs, nil
+}
+
+// IsRetriableFailure returns <false> for errors related to invalid configuration.
+func (d *ABFDescriptor) IsRetriableFailure(err error) bool {
+	return !(err == ErrABFInvalidIndex || err == ErrABFWithoutACL)
```

Comment:
IsRetriableFailure can be omitted - validation errors are automatically treated as non-retriable.